### ISSUE TITLE
IDN Domain error with cloudflare

### ISSUE THREAD

**Issue by retailify** [go-acme/lego#831]
Currently we are facing a IDN problem with lego and cloudflare. We have changed the domain to show the error and does'nt violate the customer rights.

```bash
lego --dns cloudflare --domains xn--blder-fehler-5ib.de --email mail@example.com run
lego --dns cloudflare --domains blöder-fehler.de --email mail@example.com run
```
and

I'ts the same error
```bash
2019/03/19 14:50:03 [WARN] [xn--blder-fehler-5ib.de] acme: error cleaning up: cloudflare: failed to find zonexn--blder-fehler-5ib.de.: Zone could not be found
```
We've opened a support ticket at cloudflare also.

**Comment by jmichalek132** [comment]
I hit the same issue any progress so far?

**Comment by ldez** [comment]
it seems more an error with Cloudflare than with lego.

**Comment by jmichalek132** [comment]
Did you manage to solve it?

**Comment by phenomax** [comment]
I can also reproduce the error.

**Comment by phenomax** [comment]
@retailify I can reproduce your error. The CF API Key and e-mail are correct. What did CF respond to your ticket?

**Comment by retailify** [comment]
Yes it's an error in the cloudflare api and not an issue with Lego. We does'nt onboarded the domain on cloudflare because of that error. I have not tested it again.

==>> response from cloudflare
Hi there,
DNS and CDN should work completely fine with puny domains.
This issue is only in relation to the new Registrar product.
Let me know if you have any additional questions.
For now I will mark this as solved but if you reply to the ticket it will automatically re-open.

**Comment by vlldk** [comment]
If the error is with the Cloudflare API, why did they close your ticket? Did they give you an ETA on a fix? 

It doesn't seem to be fixed, i still get the error...

**Comment by dmke** [comment]
I've encountered the problem myself, and I dug a bit into the code.

Zones names seem not to be stored in punycode form at Cloudflare, but as the Unicode (?) string the user enters when creating the zone (or its records). Using their [Go API client](https://github.com/cloudflare/cloudflare-go) we can reproduce this case:

<details><summary>(sources)</summary>

```go
package main

import (
	"flag"
	"fmt"
	"os"

	cloudflare "github.com/cloudflare/cloudflare-go"
)

var zone, token string

func main() {
	flag.StringVar(&zone, "zone", "", "zone (domain) name")
	flag.StringVar(&token, "token", "", "API token")
	flag.Parse()
	if zone == "" || token == "" {
		flag.Usage()
		os.Exit(1)
	}

	api, err := cloudflare.NewWithAPIToken(token)
	if err != nil {
		panic(err)
	}

	id, err := api.ZoneIDByName(zone)
	if err != nil {
		panic(err)
	}

	fmt.Printf("zone: %s\nid: %s\n", zone, id)
}
```

`api.ZoneIDByName` is what lego's Cloudflare provider [uses](https://github.com/go-acme/lego/blob/v3.0.2/providers/dns/cloudflare/cloudflare.go#L99).

</details>

```console
$ go run main.go -token omitted -zone with-ümlaut.com
zone: with-ümlaut.com
id: ...
$ go run main.go -token omitted -zone xn--with-mlaut-eeb.com
panic: Zone could not be found
```

The domain lego passes into [`(*providers/dns/cloudflare.DNSProvider).Present()`](https://github.com/go-acme/lego/blob/v3.0.2/providers/dns/cloudflare/cloudflare.go#L91) is already in punycode form (as a result of [`certificates.sanitizeDomain()`](https://github.com/go-acme/lego/blob/v3.0.2/certificate/certificates.go#L514) called from [`(*certificates.Certifier).Obtain()`](https://github.com/go-acme/lego/blob/v3.0.2/certificate/certificates.go#L94)).

Problem is, calling `idna.ToUnicode` (to undo the effects of `idna.ToASCII`) might not actually resolve this issue: When the zone was created as a Punycode domain in the first place, this would then break for those users... (though. I cannot say whether it is actually possible to create such a zone).

**Comment by dmke** [comment]
> I cannot say whether it is actually possible to create such a zone

At least their dashboard web app creates a zone with name `exämple.com` when I enter `xn--tost-moa.example.com`. It uses `dash.cloudflare.com/api/v4/zones` though - and their [public API](https://api.cloudflare.com/#zone-create-zone) has an ASCII constraint on the `name` attribute.

I assume the app servers behind the different URLs (`{dash,api}.cloudflare.com`) only differ in access control/rate limiting, but use the same business logic to create the zone in their databases (one would hope so ;-)).

In any case, I'd reword "calling idna.ToUnicode (to undo the effects of idna.ToASCII) might not actually resolve this issue". Calling `idna.ToUnicode` in `(*cloudflare.DNSProvider).Present()` before invoking `ZoneIDByName` should be fine.

**Comment by ldez** [comment]
Fixed by  #979
